backend/cron/service.go

```diff
@@ -0,0 +1,204 @@
+	"github.com/TBD54566975/ftl/internal/rpc"
+	"github.com/TBD54566975/ftl/internal/schema"
+	"github.com/TBD54566975/ftl/internal/slices"
+)
+
+type PullSchemaClient interface {
+	PullSchema(ctx context.Context, req *connect.Request[ftlv1.PullSchemaRequest]) (*connect.ServerStreamForClient[ftlv1.PullSchemaResponse], error)
+}
+
+type CallClient interface {
+	Call(ctx context.Context, req *connect.Request[ftlv1.CallRequest]) (*connect.Response[ftlv1.CallResponse], error)
+}
+
+type cronJob struct {
+	module  string
+	verb    *schema.Verb
+	cronmd  *schema.MetadataCronJob
+	pattern cron.Pattern
+	next    time.Time
+}
+
+func (c cronJob) String() string {
+	desc := fmt.Sprintf("%s.%s (%s)", c.module, c.verb.Name, c.pattern)
+	var next string
+	if time.Until(c.next) > 0 {
+		next = fmt.Sprintf(" (next run in %s)", time.Until(c.next))
+	}
+	return desc + next
+}
+
+// Start the cron service. Blocks until the context is cancelled.
+func Start(ctx context.Context, pullSchemaClient PullSchemaClient, verbClient CallClient) error {
+	wg, ctx := errgroup.WithContext(ctx)
+	changes := make(chan *ftlv1.PullSchemaResponse, 8)
+	// Start processing cron jobs and schema changes.
+	wg.Go(func() error {
+		return run(ctx, verbClient, changes)
+	})
+	// Start watching for schema changes.
+	wg.Go(func() error {
+		rpc.RetryStreamingServerStream(ctx, "pull-schema", backoff.Backoff{}, &ftlv1.PullSchemaRequest{}, pullSchemaClient.PullSchema, func(ctx context.Context, resp *ftlv1.PullSchemaResponse) error {
+			changes <- resp
+			return nil
+		}, rpc.AlwaysRetry())
+		return nil
+	})
+	err := wg.Wait()
+	if err != nil {
+		return fmt.Errorf("cron service stopped: %w", err)
+	}
+	return nil
+}
+
+func run(ctx context.Context, verbClient CallClient, changes chan *ftlv1.PullSchemaResponse) error {
+	logger := log.FromContext(ctx).Scope("cron")
+	// Map of cron jobs for each module.
+	cronJobs := map[string][]cronJob{}
+	// Cron jobs ordered by next execution.
+	cronQueue := []cronJob{}
+
+	logger.Debugf("Starting cron service")
+
+	for {
+		next, ok := scheduleNext(cronQueue)
+		var nextCh <-chan time.Time
+		if ok {
+			logger.Tracef("Next cron job scheduled in %s", next)
+			nextCh = time.After(next)
+		}
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("cron service stopped: %w", ctx.Err())
+
+		case resp := <-changes:
+			if err := updateCronJobs(cronJobs, resp); err != nil {
+				logger.Errorf(err, "Failed to update cron jobs")
+				continue
+			}
+			cronQueue = rebuildQueue(cronJobs)
+
+		// Execute scheduled cron job
+		case <-nextCh:
+			job := cronQueue[0]
+			logger.Debugf("Executing cron job %s", job)
+
+			nextRun, err := cron.Next(job.pattern, false)
+			if err != nil {
+				logger.Errorf(err, "Failed to calculate next run time")
+				continue
+			}
+			job.next = nextRun
+			cronQueue[0] = job
+			orderQueue(cronQueue)
+
+			if err := callCronJob(ctx, verbClient, job); err != nil {
+				logger.Errorf(err, "Failed to execute cron job")
```

Comment:
Looks like we're dropping support for retry policies on cron? We should probably make that a validation check to ensure no retry directives on cron job verbs if so.

Reply:
Or maybe just a ticket to add retry logic back in

Reply:
It seems likely we'll be moving retries to Istio.